### PR TITLE
feat(mep): Add `build_snql_query` method to replace `build_snuba_filter`

### PR DESCRIPTION
**Pull request by wedamija**
This adds in `build_snql_query`, which is intended to replace `build_snuba_filter` everywhere.
Initially, I'm just implementing this in the error and transaction entities. I'll follow up with
sessions and metrics in a later pr.

This function uses `QueryBuilder` to build the snql queries that we'll use to create alert rules.
Currently, `QueryBuilder` requires that all queries have a start/end passed since those are required
for all on demand queries that users make. Alert rules operate differently - we explicitly can't
pass a start/end, since these alerts operate on a time window, and snuba adds in the time component
every time these subscriptions run. To support this, I've added the ability to skip start/end checks
in `QueryBuilder`.

For testing, I just copied `BuildSnubaFilterTest` and converted it to check snql queries instead.
Since `build_snuba_filter` will go away soon I'm not concerned about duplication here. A bunch of
tests here are still commented out - these are all session/metric related. I'll fix those tests in
a follow up pr as well.

### CODE REVIEW

**Comment by wedamija** [comment]
One other note: To validate queries are equivalent, I ran this for each test

```
    def compare_filter_to_snql(
        self, dataset, aggregate, query, environment=None, entity_extra_fields=None
    ):
        from snuba_sdk.legacy import json_to_snql
        time_window = 3600
        entity_subscription = get_entity_subscription_for_dataset(
            dataset=dataset,
            aggregate=aggregate,
            time_window=time_window,
            extra_fields=entity_extra_fields,
        )
        snuba_filter = build_snuba_filter(
            entity_subscription,
            query,
            environment=environment,
            params={
                "organization_id": self.organization.id,
                "project_id": [self.project.id],
            },
        )
        body = {
            "project_id": self.project.id,
            "project": self.project.id,  # for SnQL SDK
            "dataset": dataset.value,
            "conditions": snuba_filter.conditions,
            "aggregations": snuba_filter.aggregations,
            "time_window": time_window,
            "resolution": 60,
            **entity_subscription.get_entity_extra_params(),
        }
        converted_filter = json_to_snql(body, entity_subscription.entity_key.value)

        snql_query = build_snql_query(
            entity_subscription,
            query,
            (self.project.id,),
            environment=environment,
            params={
                "organization_id": self.organization.id,
                "project_id": [self.project.id],
            },
        )
        print("converted_filter", converted_filter.print())
        print("snql_query", snql_query.print())
        assert False
```
This produces the queries from the old and new systems. I then manually compared them for equivalency. I couldn't automate this, since the two systems produce equivalent queries, but they're not necessarily identical (terms can be in different orders, extra parens, etc)

**Comment by barkbarkimashark** [comment]
> One other note: To validate queries are equivalent, I ran this for each test
> 
> ```
>     def compare_filter_to_snql(
>         self, dataset, aggregate, query, environment=None, entity_extra_fields=None
>     ):
>         from snuba_sdk.legacy import json_to_snql
>         time_window = 3600
>         entity_subscription = get_entity_subscription_for_dataset(
>             dataset=dataset,
>             aggregate=aggregate,
>             time_window=time_window,
>             extra_fields=entity_extra_fields,
>         )
>         snuba_filter = build_snuba_filter(
>             entity_subscription,
>             query,
>             environment=environment,
>             params={
>                 "organization_id": self.organization.id,
>                 "project_id": [self.project.id],
>             },
>         )
>         body = {
>             "project_id": self.project.id,
>             "project": self.project.id,  # for SnQL SDK
>             "dataset": dataset.value,
>             "conditions": snuba_filter.conditions,
>             "aggregations": snuba_filter.aggregations,
>             "time_window": time_window,
>             "resolution": 60,
>             **entity_subscription.get_entity_extra_params(),
>         }
>         converted_filter = json_to_snql(body, entity_subscription.entity_key.value)
> 
>         snql_query = build_snql_query(
>             entity_subscription,
>             query,
>             (self.project.id,),
>             environment=environment,
>             params={
>                 "organization_id": self.organization.id,
>                 "project_id": [self.project.id],
>             },
>         )
>         print("converted_filter", converted_filter.print())
>         print("snql_query", snql_query.print())
>         assert False
> ```
> 
> This produces the queries from the old and new systems. I then manually compared them for equivalency. I couldn't automate this, since the two systems produce equivalent queries, but they're not necessarily identical (terms can be in different orders, extra parens, etc)

Maybe, dumb suggestion, but if the two queries are effectively the same (just different ordering and extra parens), could you parse the querystring and do a character frequency count and compare the dicts - you'd obviously ignore the parens and brackets and separators.

**Comment by wedamija** [comment]
> > One other note: To validate queries are equivalent, I ran this for each test
> > ```
> >     def compare_filter_to_snql(
> >         self, dataset, aggregate, query, environment=None, entity_extra_fields=None
> >     ):
> >         from snuba_sdk.legacy import json_to_snql
> >         time_window = 3600
> >         entity_subscription = get_entity_subscription_for_dataset(
> >             dataset=dataset,
> >             aggregate=aggregate,
> >             time_window=time_window,
> >             extra_fields=entity_extra_fields,
> >         )
> >         snuba_filter = build_snuba_filter(
> >             entity_subscription,
> >             query,
> >             environment=environment,
> >             params={
> >                 "organization_id": self.organization.id,
> >                 "project_id": [self.project.id],
> >             },
> >         )
> >         body = {
> >             "project_id": self.project.id,
> >             "project": self.project.id,  # for SnQL SDK
> >             "dataset": dataset.value,
> >             "conditions": snuba_filter.conditions,
> >             "aggregations": snuba_filter.aggregations,
> >             "time_window": time_window,
> >             "resolution": 60,
> >             **entity_subscription.get_entity_extra_params(),
> >         }
> >         converted_filter = json_to_snql(body, entity_subscription.entity_key.value)
> > 
> >         snql_query = build_snql_query(
> >             entity_subscription,
> >             query,
> >             (self.project.id,),
> >             environment=environment,
> >             params={
> >                 "organization_id": self.organization.id,
> >                 "project_id": [self.project.id],
> >             },
> >         )
> >         print("converted_filter", converted_filter.print())
> >         print("snql_query", snql_query.print())
> >         assert False
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This produces the queries from the old and new systems. I then manually compared them for equivalency. I couldn't automate this, since the two systems produce equivalent queries, but they're not necessarily identical (terms can be in different orders, extra parens, etc)
> 
> Maybe dumb suggestion, but if the two queries are effectively the same (just different ordering and extra parens), could you parse the querystring and do a character frequency count and compare the dicts - you'd obviously ignore the parens and brackets and separators.
> 
> I realize we probably don't wanna write a parser to parse the query, but if we know more about the structure of the query, maybe we can use that to make the comparison more reliable.

This is a decent idea, although removing parens can change the meaning of the query when an OR gets involved. Tbh, I'm most of the way through this already, and we don't really need to keep the systems in sync long term - we'll remove the old one in a relatively short amount of time once we've switched over. So I'm happy enough to just do manual verification here.

File: src/sentry/search/events/builder.py [hunk 1]
```diff
@@ -169,8 +175,8 @@ def resolve_time_conditions(self) -> None:
         ), "Both start and end params must be datetime objects"
 
         # Strip timezone, which are ignored and assumed UTC to match filtering
-        self.start: datetime = self.params["start"].replace(tzinfo=timezone.utc)
-        self.end: datetime = self.params["end"].replace(tzinfo=timezone.utc)
```
Comment (barkbarkimashark) [inline src/sentry/search/events/builder.py:173]: Why'd you remove this?

**Review by barkbarkimashark — APPROVED** [review]
Not a subject matter expert on Snuba querying, but I looked over the code and I don't see anything glaringly wrong. 

* `build_snql_query` (which uses QueryBuilder) to replace `build_snuba_filter`. These should be functionally equivalent. I assume someone has validated that assumption, so not really in scope for this. 
* `QueryBuilder` seems to require `start` and `end`, but it looks like you introduced a bypass to skip those conditions.
* porting a bunch of tests for `build_snuba_filter`

File: src/sentry/search/events/builder.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/search/events/builder.py:173]: I added the definition in `__init__` instead, and mypy was complaining that it these variables were already defined.

File: src/sentry/search/events/builder.py [hunk 2]
```diff
@@ -109,6 +109,7 @@ def __init__(
         turbo: bool = False,
         sample_rate: Optional[float] = None,
         equation_config: Optional[Dict[str, bool]] = None,
+        skip_time_conditions: bool = False,
```
Comment (wmak) [inline src/sentry/search/events/builder.py:112]: Could we rename the variable, or add a comment explaining that this is specifically for alerts?

**Review by wmak — APPROVED** [review]
Minor comment about variable naming otherwise LGTM :+1:

**Pull request merged**